azkaban-common/src/main/java/azkaban/executor/container/watch/ContainerStatusMetricsListener.java

```diff
@@ -0,0 +1,186 @@
+      Logger.getLogger(ContainerStatusMetricsListener.class);
+  private static final int DEFAULT_EVENT_CACHE_MAX_ENTRIES = 50000;
+  private static final int SHUTDOWN_TERMINATION_TIMEOUT_SECONDS = 5;
+
+  private final ContainerizationMetrics containerizationMetrics;
+
+  // Since cached data is key-value pair of pod name and pod status from event metadata,
+  // each data size is expected to be no more than 100 bytes, maximum cache size will
+  // be 100 bytes * DEFAULT_EVENT_CACHE_MAX_ENTRIES ~ 5 mb
+  private final Cache<String, AzPodStatus> podStatusCache;
+
+  private final ExecutorService executor;
+
+  @Inject
+  public ContainerStatusMetricsListener(final ContainerizationMetrics containerizationMetrics) {
+    this.containerizationMetrics = containerizationMetrics;
+
+    this.executor = Executors.newSingleThreadExecutor(
+        new ThreadFactoryBuilder().setNameFormat("azk-container-metrics-pool-%d").build());
+
+    this.podStatusCache = CacheBuilder.newBuilder()
+        .maximumSize(DEFAULT_EVENT_CACHE_MAX_ENTRIES)
+        .recordStats()
+        .build();
+  }
+
+  /**
+   * Checks whether the {@link AzPodStatus} of a new event is the same as the one already cached.
+   *
+   * @param event pod watch event
+   * @return true if the status is different from cached value, false otherwise
+   */
+  protected boolean isUpdatedPodStatusDistinct(final AzPodStatusMetadata event) {
+    AzPodStatus currentStatus = AzPodStatus.AZ_POD_UNSET;
+    currentStatus = podStatusCache.getIfPresent(event.getPodName());
+
+    boolean shouldSkip = currentStatus == event.getAzPodStatus();
+    if (shouldSkip) {
+      logger.info(format("Event pod status is same as current %s, for pod %s."
+          +" Any updates will be skipped.", currentStatus, event.getPodName()));
+    }
+    return !shouldSkip;
+  }
+
+  // Update the cache with the given event.
+  protected void updatePodStatus(final AzPodStatusMetadata event) {
+    podStatusCache.put(event.getPodName(), event.getAzPodStatus());
+    logger.debug(format("Updated status to %s, for pod %s", event.getAzPodStatus(), event.getPodName()));
+  }
+
+  /**
+   * Validate and process event data to emit AzPodStatus metrics
+   * @param event
+   */
+  private void validateAndProcess(final AzPodStatusMetadata event) {
+    requireNonNull(event, "event must be non-null");
+    if (!event.getFlowPodMetadata().isPresent() || !isUpdatedPodStatusDistinct(event)) {
+      return;
+    }
+    // Update AzPodStatus metrics for the flow-pod respectively
+    switch (event.getAzPodStatus()) {
+      case AZ_POD_REQUESTED:
+        containerizationMetrics.markPodRequested();
+        break;
+      case AZ_POD_SCHEDULED:
+        containerizationMetrics.markPodScheduled();
+        break;
+      case AZ_POD_INIT_CONTAINERS_RUNNING:
+        containerizationMetrics.markInitContainerRunning();
+        break;
+      case AZ_POD_APP_CONTAINERS_STARTING:
+        containerizationMetrics.markAppContainerStarting();
```

Comment:
@sshardool Have we considered monitoring the time it takes for `containers` and `app container` to init? If yes, I wonder how can we do that? When do we say the initialization process started/finished?

Reply:
Yes, as per the plan a subsequent set of changes (from @hqin2016) will add metrics for time spent in the individual states defined in `AzPodStatus`. 
These states are intended to reflect the `init` time with sufficient (but not excessive) granularity. Initialization of the pod will be split into a few key states
- Time taken for scheduling the pod (by the K8S ApiServer)
- Time taken for running the init-containers
- Time taken for launching the application-container(s)
Metrics for these will reflect the `pod init` times. There will be room for making these even more granular if we see an operational need for it in future.